pkg/cloud/aws/actuators/machine/actuator_test.go

```diff
@@ -498,3 +506,377 @@ func TestImmutableStateChange(t *testing.T) {
+			expectedIsNodeJoin:        false,
+			expectedError:             fmt.Errorf("Unknown value %q for label `set` on machine %q", "random", "random-0"),
+		},
+		{
+			name:                      "should return error when called with invalid scope",
+			inputScope:                getInvalidMachineScope(t),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        false,
+			expectedError:             fmt.Errorf("failed to create machine scope for machine %q in namespace %q: failed to generate new scope from nil cluster", "master-0", "awesome-ns"),
+		},
+		{
+			name:                      "should return error for no `set` label",
+			inputScope:                getNoLabelMachineScope(t, nil),
+			inputControlplaneMachines: nil,
+			expectedIsNodeJoin:        false,
+			expectedError:             fmt.Errorf("Unknown value %q for label `set` on machine %q", "", "nolabel-0"),
+		},
+		{
+			name:                      "should not join first controlplane machine",
+			inputScope:                getControlplaneMachineScope(t, nil),
+			inputControlplaneMachines: nil,
+			expectedIsNodeJoin:        false,
+			expectedError:             nil,
+		},
+		{
+			name: "should not join controlplane machine when no controlplane machines exist",
+			inputScope: getControlplaneMachineScope(
+				t,
+				getMockEC2APIDescribeInstancesNotFound([]string{"master-0", "master-1"}, mockCtrl),
+			),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        false,
+			expectedError:             nil,
+		},
+		{
+			name: "should return error when unable to verify controlplane machine existence",
+			inputScope: getControlplaneMachineScope(
+				t,
+				getMockEC2APIDescribeInstancesFail([]string{"master-0", "master-1"}, mockCtrl),
+			),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        false,
+			expectedError: fmt.Errorf(
+				`failed to verify existence of machine "master-0" in namespace "awesome-ns": failed to lookup machine "master-0": failed to describe instances by tags: mock API failure`),
+		},
+		{
+			name: "should join controlplane machine when other controlplane machine exists",
+			inputScope: getControlplaneMachineScope(
+				t,
+				getMockEC2APIDescribeInstancesPass([]string{"master-0", "master-1"}, mockCtrl),
+			),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        true,
+			expectedError:             nil,
+		},
+		{
+			name: "should join controlplane machine when first controlplane machine doesn't exist but second controlplane machine exists",
+			inputScope: getControlplaneMachineScope(
+				t,
+				getMockEC2APIDescribeInstancesNotFoundAndPass([]string{"master-0", "master-1"}, mockCtrl),
+			),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        true,
+			expectedError:             nil,
+		},
+	}
+
+	for _, tc := range testCases {
```

Comment:
this needs to run each test case as a subtest using:

```
for _, tc := range testCases {
    t.Run(tc.name, func(t *testing.T) {
        ...everything else
    }
}
```

Reply:
Will do. Is the benefit here just to let the framework run the tests or are there other benefits or running it this way?

Reply:
benefit is that a single subtest can fail while others continue to run. We get good output about which test case in the test table actually failed. Without it we just know this method failed but not which case.